rueidisprob/bloomfilter.go

```diff
@@ -0,0 +1,352 @@
+	Count(ctx context.Context) (uint, error)
+}
+
+type bloomFilter struct {
+	client rueidis.Client
+
+	// name is the name of the Bloom filter.
+	// It is used as a key in the Redis.
+	name string
+
+	// counter is the name of the counter.
+	counter string
+
+	// hashIterations is the number of hash functions to use.
+	hashIterations      uint
+	hashIterationString string
+
+	// size is the number of bits to use.
+	size uint
+
+	addMultiScript *rueidis.Lua
+	addMultiKeys   []string
+
+	existsMultiScript *rueidis.Lua
+	existsMultiKeys   []string
+}
+
+// NewBloomFilter creates a new Bloom filter.
+// NOTE: 'name:c' is used as a counter key in the Redis
+// to keep track of the number of items in the Bloom filter for Count method.
+func NewBloomFilter(
+	client rueidis.Client,
+	name string,
+	expectedNumberOfItems uint,
+	falsePositiveRate float64,
+) (BloomFilter, error) {
+	if len(name) == 0 {
+		return nil, ErrEmptyName
+	}
+
+	if falsePositiveRate <= 0 {
+		return nil, ErrFalsePositiveRateLessThanEqualZero
+	}
+	if falsePositiveRate > 1 {
+		return nil, ErrFalsePositiveRateGreaterThanOne
+	}
+
+	size := numberOfBits(expectedNumberOfItems, falsePositiveRate)
+	if size == 0 {
+		return nil, ErrBitsSizeZero
+	}
+	if size > maxSize {
+		return nil, ErrBitsSizeTooLarge
+	}
+	hashIterations := numberOfHashFunctions(size, expectedNumberOfItems)
+
+	// NOTE: https://redis.io/docs/reference/cluster-spec/#hash-tags
+	bfName := "{" + name + "}"
+	counterName := bfName + ":c"
+	return &bloomFilter{
+		client:              client,
+		name:                bfName,
+		counter:             counterName,
+		hashIterations:      hashIterations,
+		hashIterationString: strconv.FormatUint(uint64(hashIterations), 10),
+		size:                size,
+		addMultiScript:      rueidis.NewLuaScript(addMultiScript),
+		addMultiKeys:        []string{bfName, counterName},
+		existsMultiScript:   rueidis.NewLuaScript(existsMultiScript),
+		existsMultiKeys:     []string{bfName},
+	}, nil
+}
+
+func numberOfBits(n uint, r float64) uint {
+	return uint(math.Ceil(-float64(n) * math.Log(r) / math.Pow(math.Log(2), 2)))
+}
+
+func numberOfHashFunctions(s uint, n uint) uint {
+	return uint(math.Round(float64(s) / float64(n) * math.Log(2)))
+}
+
+func (c *bloomFilter) Add(ctx context.Context, key string) error {
+	return c.AddMulti(ctx, []string{key})
+}
+
+func (c *bloomFilter) AddMulti(ctx context.Context, keys []string) error {
+	if len(keys) == 0 {
+		return nil
+	}
+
+	indexes := c.indexes(keys)
+
+	args := make([]string, 0, len(indexes)+1)
+	args = append(args, c.hashIterationString)
+	args = append(args, indexes...)
+
+	resp := c.addMultiScript.Exec(ctx, c.client, c.addMultiKeys, args)
+	if resp.Error() != nil {
+		return resp.Error()
+	}
+
+	return nil
+}
+
+func (c *bloomFilter) indexes(keys []string) []string {
+	allIndexes := make([]string, 0, len(keys)*int(c.hashIterations))
+	for _, key := range keys {
+		indices := indexes([]byte(key), c.hashIterations, c.size)
```

Comment:
Can we reuse the `indices` for every key to save some allocations? Or how about letting the `indexes` function write to the `allIndexes` directly?

Reply:
[56fc836](https://github.com/redis/rueidis/pull/493/commits/56fc836c835bd589c7bbcc61d063e85835f59cb2)